### PR TITLE
File input multi2

### LINKED ISSUE #9727: [FEATURE] Add multi-file selection to FileInput widget
**The ISSUE**:
The current version (bokeh 1.4, branch master) of the FileInput widget of bokeh only allows a single file to be selected by the user.
The underlying <input type='file'> item in the browser comes with the option/property "multiple" in JS or HTML5, which is hard-coded to "False" in the bokeh-code.

**The Problem**:
For data analysis project in my laboratory I need to upload many tenth of files at a time, which is not manageable by single upload. 
Hence the request (**and a related PR**) to add a working "multiple"-property and the respective option of multi-file selection to the FileInput widget.

**Solution**:
Adapt/extend the code of the currently implemented FileInput to support a "multiple"-property with the (switchable) capability to support the multi-file selection and provide the returned results as arrays instead of scalar values.

**Background**:
The issue was [pre-discussed here in community support](https://discourse.bokeh.org/t/why-is-bokeh-fileinput-limited-to-single-selection/4778) and [here in the develeopment-forum](https://discourse.bokeh.org/t/callback-triggering-broken-after-change-from-string-to-array-in-fileinput-widget/4797).

**possible Alternative**:
The implementation of a seperate widget (MultiFileInput) was considered but rejected because of the inconsistency with usual HTML or JS input widgets of similar functionality.

### PR DESCRIPTION
**Pull request by bruns01**
- [ ] issues: fixes #9727
**This PR replaces PR #9728**

The PR amends the capability of multi-file selection to bokeh's FileInput widget.
It adapts the property types of class FileInput in
- inputs.py

and adapts the help-texts.
It extends the TS-code in
- file_input.ts

to ensure complete file reading/upload and return of either scalars (single file API) or Arrays (multi file API).

### CODE REVIEW

File: bokeh/models/widgets/inputs.py [hunk 1]
```diff
@@ -89,48 +89,49 @@ def coerce_value(cls, val):
...
-
     `video/*`:
         all video files are pickable
-
     `image/*`:
         all image files are pickable
-
     `<media type>`:
         A valid `IANA Media Type`_, with no parameters.
-
```
Comment (bryevdv) [inline bokeh/models/widgets/inputs.py]: @bruns01 Sphinx is very picky and fussy. The whitespace lines here is actually meaningful and necessary for Sphinx to generate the rendered HTML docs correctly.

**Comment by bryevdv** [comment]
@bruns01 This is is much cleaner, thank you! I left a comment about the docs build issue, and otherwise I think there is just one more trailing whitespace on line 54 to get the build green.

File: bokeh/models/widgets/inputs.py [hunk 2]
```diff
@@ -89,48 +89,49 @@ def coerce_value(cls, val):
...
-    .. note::
-        The full file path is not included since browsers will not provide
-        access to that information for security reasons.
-    """)
+    
+    value = Either(String, List(String), default='', readonly=True, help='''
+    if multiple=False or ommitted, base64-encoded string of the contents of the 
+    selected file,
+    if multiple=True or 'Multiple' a list of base64 encoded strings of the contents
```
Comment (bryevdv) [inline bokeh/models/widgets/inputs.py]: comment about `"Multiple"` setting value can be removed (here and below). I might also suggest verbiage with an initial short sentence, then further explanation,e.g. 

```
    value = Either(String, List(String), default='', readonly=True, help='''
    The base64-enconded contents of the file or files that were loaded.

    If `mulitiple` is set to False (default), this value is a single string with the contents
    of the single file that was chosen.

    If `multiple` is True, this value is a list of strings, each contenting the contents of 
    one of the multiple files that were chosen. 
    ''')
``` 

And similar below, what do you think?

Comment (bruns01) [reply bokeh/models/widgets/inputs.py]: Applied in the new commit to all three items (value, filename, mime-type).
I changed "contenting" to "containing", although I'm not native speaker, it seemed better.

File: bokeh/models/widgets/inputs.py (hunk 1 above)
Comment (bruns01) [reply bokeh/models/widgets/inputs.py]: Reapplied the empty lines.
(I don't remember to have them intentionally removed in the first place (???).)

File: bokehjs/src/lib/models/widgets/file_input.ts [hunk 3]
```diff
@@ -16,43 +16,59 @@ export class FileInputView extends WidgetView {
...
+    const filename: string[] = []
+    const mime_type: string[] = []
+    let i: number
+
+    for (i=0; i<files.length; i++){
+      filename.push(files[i].name)
+      const content = await this.readfile(files[i])
+      const file_arr = content.split(",")
+      value.push(file_arr[1])
+      mime_type.push(file_arr[0].split(":")[1].split(";")[0])
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/file_input.ts]: This code needs to be refactored in a way that we have at least a vague understanding what it does (or alternatively, if you have to change it, let's say in a year, then you will be able to do it without any problems). So, starting from the beginning, `content` variable isn't really file's content, but a data URL. A better name would be e.g. `data_url`. Then all the further text splitting makes finally some sense to the reader, as we can infer that you are parsing said data URL. Then you extract URL's declaration and actual content. Again, `file_arr` isn't the best name, because it doesn't tell us anything about the contents of this variable, except for it being allegedly an array. We know, from data URL specification, that it contains exactly one `,`, thus we get a 2-tuple a result. Based on this knowledge, one could use `split(",", 2)` to improve performance on this line. Instead of introducing `file_arr`, it would be better to immediately unpack the result as `[declaration, content]`, resulting in code like this:

```ts
const data_url = await this.readfile(files[i])
const [declaration, content] = data_url.split(",", 2)
value.push(content)
mime_type.push(declaration.split(":")[1].split(";")[0])
```
Much better, though the last line is still a bit of a sore. That's just one approach. Alternatively you could use a regex to split the data URL in one go:

```ts
const data_url = await this.readfile(files[i])
const [, mime, , content] = data_url.split(/[:;,]/, 4)
value.push(content)
mime_type.push(mime)
```
Very readable, but performance may suffer (likely not if splitting limit is used, because the actual content isn't going to be scanned by `split()` method). It's up to you as the author to decide which is the best approach, considering criteria like correctness, readability (maintainability) and performance.

Comment (bruns01) [reply bokehjs/src/lib/models/widgets/file_input.ts]: Followed your advice and chose the later option.
The only change is I called "content" "data" instead, as it is the "data" part of the "data_url" in my view. 
Hope I found all the missing and excessive white spaces, too.
I tried to install and use tslint locally but had no success on that (so far).

File: bokeh/models/widgets/inputs.py (hunk 2 above)
Comment (bryevdv) [reply bokeh/models/widgets/inputs.py]: Yes, that was a typo on my part, thanks

File: bokehjs/src/lib/models/widgets/file_input.ts [hunk 4]
```diff
@@ -16,43 +16,59 @@ export class FileInputView extends WidgetView {
...
     this.dialogEl.disabled = this.model.disabled
   }
 
-  load_file(e: any): void {
-    const reader = new FileReader()
-    this.model.filename = e.target.files[0].name
-    reader.onload = (e) => this.file(e)
-    reader.readAsDataURL(e.target.files[0])
+  async load_files(e: any): Promise<void> {
+    const files = e.target.files
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/file_input.ts]: A lot of this code was originally written without proper care, but with the extent of changes in this PR, let's do this properly this time and actually fill in those gaps. When `e` is of type `any` then all its properties are of type `any` and so on, so we are back with dynamically typed JavaScript, and that's the last we want. `e` comes from `onchange` callback, thus we can only use the not so specific `Event` type. Though not super useful, it will already show a flaw with this code, mainly `e.target` can be `null`. You could deal with this, but it would be even better to pass `files` (of type `FileList`) directly to this method, especially given that you don't use `e` for any other purpose:
```ts
async load_files(files: FileList): Promise<void> { ... }
```
Given that we really don't rely on the event target mechanism altogether, we can use `this.dialogEl.files` directly:
```ts
this.dialogEl.onchange = async () => {
  const {files} = this.dialogEl
  if (files != null) {
    await this.load_files(files)
  }
}
```
`files` can be `null`, so we have to deal with this. I left `async` and `await`, but those aren't strictly necessary in this context, though it doesn't hurt to leave them either.

File: bokehjs/src/lib/models/widgets/file_input.ts [hunk 5]
```diff
@@ -16,43 +16,59 @@ export class FileInputView extends WidgetView {
...
-    const file_arr = file.split(",")
-
-    const content = file_arr[1]
-    const header = file_arr[0].split(":")[1].split(";")[0]
-
-    this.model.value = content
-    this.model.mime_type = header
-
+  readfile(file: any): Promise<string> {
+    return new Promise<any>((resolve) => {
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/file_input.ts]: I will leave this one as an exercise. Let me know if you have any issues/questions with this one.

File: bokehjs/src/lib/models/widgets/file_input.ts (hunk 4 above)
Comment (bruns01) [reply bokehjs/src/lib/models/widgets/file_input.ts]: Please, do not impute "carelessness" if the worker (me) notified you about a huge lack of competence and experience in TS at the very start of the comittements.
Anyway, thank you for the patience and proposals, I tried to follow with my latest commit.

File: bokehjs/src/lib/models/widgets/file_input.ts (hunk 5 above)
Comment (bruns01) [reply bokehjs/src/lib/models/widgets/file_input.ts]: My solution below gets rid of "any"-type but needs a String() cast in order to satisfy the checks of the compiler. Although _e.target.result_ should contain a string because of _readAsDataURL()_ the compiler complains about incompatible types if I remove the cast (???). Apparently there are things that I still do not understand.
```
  readfile(file: File): Promise<string> {
    return new Promise<string>((resolve) => {
      const reader = new FileReader()
      reader.onload = (e) => {
        if (e.target) {
            resolve(e.target.result ? String(e.target.result) : "")
        }
        else {resolve("")}
      }
      reader.readAsDataURL(file)
    })
  }
```

File: bokehjs/src/lib/models/widgets/file_input.ts (hunk 4 above)
Comment (bryevdv) [reply bokehjs/src/lib/models/widgets/file_input.ts]: @bruns01 @mattpap is referring to the original state of the code from before your PR, and there is indeed lots of old technical debt we can clean up (which is why PRs like his are a very nice opportunity!)

Comment (bruns01) [reply bokehjs/src/lib/models/widgets/file_input.ts]: @bryevdv  Thanks, for that clarification!

File: bokehjs/src/lib/models/widgets/file_input.ts [hunk 6]
```diff
@@ -16,43 +16,69 @@ export class FileInputView extends WidgetView {
...
-
+  readfile(file: File): Promise<string> {
+    return new Promise<string>((resolve) => {
+      const reader = new FileReader()
+      reader.onload = (e) => {
+        if (e.target) {
+          resolve(e.target.result ? String(e.target.result) : "")
+        } else {
+          resolve("")
+        }
```
Comment (mattpap) [inline bokehjs/src/lib/models/widgets/file_input.ts]: This block of code still needs some refinement. In general, `e.target.result` can be a `string` or an `ArrayBuffer` at `String()` call. However, `String()` applied to an `ArrayBuffer` gives a result far from what we would expected:
```ts
$ node
Welcome to Node.js v12.14.1.
Type ".help" for more information.
> x = new ArrayBuffer(10)
ArrayBuffer {
  [Uint8Contents]: <00 00 00 00 00 00 00 00 00 00>,
  byteLength: 10
}
> String(x)
'[object ArrayBuffer]'
```
However, we don't have to deal with `ArrayBuffer` at all, because using `readAsDataURL()` guarantees that `result` will be only of type `string | null`. Unfortunately encoding this at type-level is hard or perhaps not possible at all. So, my suggestion is to cast `result as string` to help the type system and not use otherwise incorrectly behaving `String()`.

Additionally, returning an empty string on failure will result in a failure on new line 44, where we assume a data URL. The empty string on line 66 can be eliminated easily, because we can assume `e.target == reader` and remove the if statement altogether. Or alternatively we could use `e.target!` type-level construct, telling the type-checker that we know `null` can't appear, but we can't express this by better and safer means. The remaining question is whether `result` can really be `null` at this point. After consulting FileReader's specification, I'm not sure what to expect. So, we have two options. Either carry this `null` around, which will likely mean that any errors will be silently ignored. Alternatively, we can reject the promise with an error. This is the solution I'm leaning on. 

In summary I would proceed with the following code:
```ts
    return new Promise<string>((resolve, reject) => {
      const reader = new FileReader()
      reader.onload = () => {
        const {result} = reader
        if (result != null) {
          resolve(result as string)
        } else {
          reject(reader.error ?? new Error(`unable to read '${file.name}'`))
        }
      }
      reader.readAsDataURL(file)
    })
```

Comment (bruns01) [reply bokehjs/src/lib/models/widgets/file_input.ts]: Thank you!
I implemented this in a further commit (copy/paste because such refinements are beyond my skills). 
I tested to trigger a reject() by selecting/uploading a dead soft-link without a target-file but this resulted merely in "no file selected". Selecting three files with one being the dead soft link resulted in "two files selected". So no error under this condition. (While the shell informs "no such file or directory" if I want to "> less dead_link.txt".)

**Comment by bruns01** [comment]
I saw that js-test failed and looked at the details. However, I have no clue what causes the failure!
There is no indication for me related to the committed file?!

**Comment by mattpap** [comment]
It's just chromium failing to start. It happens sometimes for not good reason.

**Comment by bruns01** [comment]
Is there anything left to do for me, now. Or can I relax and wait for the merge ;^) ?

**Comment by mattpap** [comment]
> Is there anything left to do for me, now. Or can I relax and wait for the merge ;^) ?

Looks good now. We are now waiting for the 2.0 release to happen, before we get to merging new PRs.
